runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -3859,16 +3875,94 @@ inline
 void indicate_done2(int caller, rf_done_t *ack)
 {
   static rf_done_t done = 1;
-  do_remote_put(&done, caller, ack, sizeof(done), &gnr_mreg_map[caller],
-                may_proxy_false);
+  send_polling_response(&done, caller, ack, sizeof(done));
 }
 
 
 static
-inline
-void indicate_done(fork_base_info_t* b)
+void send_polling_response(void* src_addr, c_nodeid_t locale, void* tgt_addr,
+                           size_t size)
 {
-  indicate_done2(b->caller, b->rf_done);
+  gni_post_descriptor_t* post_desc;
+
+  //
+  // If we're not the polling task, send the response the regular way.
+  // Thus we only do asynchronous completion processing on polling task
+  // response transactions initiated on its firmly bound comm domain.
+  // This gets nearly all the benefit of not waiting for completions,
+  // while avoiding concurrency control entirely.
+  //
+  if (cd == NULL || !cd->firmly_bound) {
+    do_remote_put(src_addr, locale, tgt_addr, size, &gnr_mreg_map[locale],
+                  may_proxy_false);
+    return;
+  }
+
+  DBG_P_LP(DBGF_GETPUT, "SendPollResp %p -> %d:%p (%#zx)",
+           src_addr, (int) locale, tgt_addr, size);
+
+  {
+    //
+    // This pool of GNI Post descriptors is for the polling task to use
+    // for request responses (buffer releases and "done" indicators).
+    // To improve throughput, we don't wait for the remote completions
+    // when we send these.  Instead, we collect the CQ completion events
+    // later, as they come in.  We can have as many request responses in
+    // flight simultaneously as we have Post descriptors in the pool.
+    // Small-scale runs on Aries-based systems show the polling task can
+    // do about 3 minimum-cost requests ("fast" remote forks, etc.) in
+    // the minimum network latency of ~1 usec.  The pool size we use
+    // here is enough larger that it will cover the max network latency
+    // of ~1.5 usec.
+    //
+    // The post_id members of these are used as alloc/free indicators,
+    // with 0 meaning "free".
+    //
+#define PPDESCS_CNT 8  // must be a power of 2; see PPDI_NEXT()
+
+#define PPDI_NEXT(_i) (((_i) + 1) & (PPDESCS_CNT - 1))
+
+    static gni_post_descriptor_t polling_post_descs[PPDESCS_CNT] = { 0 };
+    static int last_ppdi = 0;
+    int ppdi;
+
+    ppdi = last_ppdi;
+    while (polling_post_descs[ppdi].post_id != 0) {
+      if ((ppdi = PPDI_NEXT(ppdi)) == last_ppdi) {
+        local_yield();
+        consume_all_outstanding_cq_events(cd_idx);
+      }
+    }
+
+    last_ppdi = PPDI_NEXT(ppdi);
+
+    post_desc = &polling_post_descs[ppdi];
+    post_desc->post_id = 1;  // mark as allocated
+
+#undef PPDESCS_CNT
+#undef PPDI_NEXT
+  }
+
+  //
+  // Fill in the POST descriptor.
+  //
+  post_desc->type            = GNI_POST_FMA_PUT;
+  post_desc->cq_mode         = GNI_CQMODE_GLOBAL_EVENT;
+  post_desc->dlvr_mode       = GNI_DLVMODE_PERFORMANCE;
+  post_desc->rdma_mode       = 0;
+  post_desc->src_cq_hndl     = 0;
```

Comment:
I'm thinking/guessing/assuming that these first five fields aren't modified by ``post_fma()`` or anything else?  Would there be an advantage to getting them all set once up front, and not setting them again each time through?

Reply:
The documentation doesn't say whether uGNI might modify these or not, so I don't think we can rely on it not doing so.  I doubt any benefit would be noticeable in practice anyway, since the total number of stores needed to initiate a network transaction is large and some of them are expensive (the ones to the memory-mapped FMA descriptor and window).  And the code is a bit more clear with these explicit at the point of the post.